src/ophyd_async/core/mock_signal_utils.py

```diff
@@ -0,0 +1,183 @@
+    # global attrbute here.
+    require_all_consumed: bool = False
+
+    def __init__(
+        self,
+        signal: Signal,
+        *values: List[Any],
+        repeat: bool = False,
+        require_all_consumed: bool = False,
+    ):
+        if require_all_consumed and repeat:
+            raise ValueError(
+                "`require_all_consumed` and `repeat` cannot both be True, "
+                "as the cyclic values will never be considered fully consumed."
+            )
+        self.signal = signal
+        self.values = values
+        self.require_all_consumed = require_all_consumed
+        self.index = 0
+
+        maybe_repeat = itertools.cycle if repeat else iter
+        self.iterator = enumerate(maybe_repeat(values), start=1)
+
+    def __iter__(self):
+        return self
+
+    def __next__(self):
+        # Will propogate StopIteration
+        self.index, next_value = next(self.iterator)
+        set_mock_value(self.signal, next_value)
+        return next_value
+
+    def __del__(self):
+        if self.require_all_consumed and self.index != len(self.values):
+            raise AssertionError("Not all values have been consumed.")
+
+
+def set_mock_values(
+    signal: Signal,
+    *values: List[Any],
+    repeat: bool = False,
+    require_all_consumed: bool = False,
+) -> Iterator[Any]:
+    """Iterator to set a signal to a sequence of values, optionally repeating the
+    sequence.
+
+    Parameters
+    ----------
+    signal:
+        A signal with a `MockSignalBackend` backend.
+    values:
+        The values to set the signal to, on each iteration the value will be set.
+    repeat:
+        If True, the values will be repeated indefinitely, starting over once exhausted.
+    require_all_consumed:
+        If True, an AssertionError will be raised if the iterator is deleted before
+        all values have been consumed. Not compatible with `repeat=True`.
+
+    Notes
+    -----
+    Example usage::
+
+           for value_set in set_mock_values(signal, 1, 2, 3, repeat=True):
+                # do something
+
+            cm = set_mock_values(signal, 1, 2, 3, require_all_consumed=True):
+            next(cm)
+            # do something
+    """
+
+    return _SetValuesIterator(
+        signal,
+        *values,
+        repeat=repeat,
+        require_all_consumed=require_all_consumed,
+    )
+
+
+# linting isn't smart enought to realize @contextmanager will give use a
+# ContextManager[None]
+@contextmanager
+def set_callback_on_mock_put(
+    signal: Signal, callback: Callable[[T], None]
+) -> Generator[None, None, None]:
+    """A context manager to call a callback when a backend is put to.
+
+    Parameters
+    ----------
+    signal:
+        A signal with a `MockSignalBackend` backend.
+    callback:
+        The callback to call when the backend is put to during the context.
+    """
+    backend = _get_mock_signal_backend(signal)
+
+    # Save the original put method
+    original_put = backend.put
```

Comment:
Could we do `backend.mock.put.side_effect = callback` instead?

Reply:
You don't think it's a little nicer to have a dedicated callback in `MockSignalBackend`?

``` python
    def set_callback(self, callback: Optional[ReadingValueCallback[T]]) -> None:
        if callback:
            assert not self.callback, "Cannot set a callback when one is already set"
            reading: Reading = self.converter.reading(
                self._value, self._timestamp, self._severity
            )
            callback(reading, self._value)
        self.callback = callback
```

This allows for easier reading metadata in tests:
``` python
    assert calls == [
        (
            {
                "alarm_severity": 0,
                "timestamp": ANY,
                "value": 0.0,
            },
            0.0,
        ),
    ]
```